Neos.Flow/Configuration/Settings.Http.yaml

```diff
@@ -16,13 +16,8 @@ Neos:
       #
       applicationToken: 'MinorVersion'
 
-      # Defines an explicit base URI that should be used. This affects
-      # resource management, routing and all other parts accessing
-      # "Http\Request->getBaseUri()".
-      #
-      # If this setting is NULL, the base URI will be detected from the current
-      # request (default).
-      baseUri: NULL
+      # Deprecated with Flow 8.3
```

Comment:
How will current use cases of this be handled afterwards, I see no suitable replacement for eg. resource uris or command line uri building (especially outside of a neos context).

Reply:
related: https://github.com/neos/flow-development-collection/issues/2157

for cli you can just do:

```
$actionUriBuilderFactory->createFromBaseUri(new Uri("https://mybaseuri.de")); // or inject a setting etc
```

which is imo much better than what we can do with the current uribuilder / baseuri (the base uri is just often useless in a multisite, or when im using another host so it should be given as cli param)

cant say much about the other use cases ...

Reply:
IIRC the Resource Manager uses the current HTTP Request. For CLI usage instead of configuring a global base URI, this can (and IMO should) be configured at the corresponding ResourceTarget (or Collection?) instead.
But we should document and test this.
In general I think we should never add a `@deprecated` annotation without describing/linking to alternatives and/or explaining why this concept is considered bad

Reply:
Right, Resources use both in a fallback fahion (`\Neos\Flow\Http\BaseUriProvider::getConfiguredBaseUriOrFallbackToCurrentRequest`) 

I guess directly configurating the baseUri on the target(s) is fine, might become a bit cumbersome if you have multiple, but with ENV that should do.

I am still unsure how the CLI usecase for complex stuff (eg. rendering fusion with links or mails and such) is to be handled?

Reply:
`resolveAssetUri` could take in the baseUri or current request …

The linking service is generally due for a refactor after this pr 😄 … we can plan any breaking changes for 9.0 then …

Reply:
But if we require a second argument something like the `Neos.Link` helper becomes less sexy:

```
${Neos.Link.resolveAssetUri(uri)}
# new
${Neos.Link.resolveAssetUri(uri, request)}
```

on the other hand, the `Neos.Link.resolveNodeUri` also requires already the controllerContext ... so ...

Reply:
well i discussed this dilemma with @sorenmalling and i get know why we can pass the baseUri or request that simple to `Neos.Link.resolveAssetUri` as there ResourceTargetInterface would need to be changed too and for non file system targets, its completely irrelevant what the base uri might be (cloud storage) so it might be odd to pass the base uri along.

I guess we are still stuck with some global singleton baseUri configuration which can be written to...

idea (pseudo code):


in the cli handler
```php
// set the baseuri via env or cli argument
$baseUri = getEnv("FLOW_BASE_URI") || $cliArguments["baseUri"];
$this->baseUriConfiguration->withBaseUri(new Uri($baseUri), fn () => $this->processCliRequest());
```

in the web handler
```php
// base uri inferred from request
$this->baseUriConfiguration->withBaseUri(
    RequestInformationHelper::baseUriFromRequest($httpRequest),
    fn () => $this->processRequest($httpRequest)
);
```


in my custom code, when i want to temporary change the base uri:
```php

$output = [];
foreach($siteUri as $uri) {
   $output[] = $this->baseUriConfiguration->withBaseUri($uri, fn () => $this->fusionRenderNode($node));
}

```

Reply:
It compose the baseuri of the the target options with the current host … and an absolute uri (with host) is required for emails fx…

Reply:
Just briefly discussed with @mhsdesign and after thinking about it a couple of minutes longer I think that we don't have to implement any further `baseUri` replacement _for now_.

In the `@deprecated` annotation and the upgrade guides we should explain how you are supposed to work around the globally configured base URI.

Some further thoughts and ideas:

### MVC URIs

To create absolute URIs for CLI or other "out of band rendering", the `ActionUriBuilderFactory::createFromBaseUri()` can be used.

But we have to think about solutions for Fusion (and maybe even Fluid?).
E.g. the `Neos.Fusion:ActionUri` Fusion prototype currently uses the action request to resolve default package, controller, ... and base URI.

A separate `baseUri` prop could work:

```
uri = Neos.Fusion:ActionUri {
  baseUri = ${Configuration.setting('Some.Package:someBaseUri')}
  package = 'Some.Package'
  controller = 'Standard'
  action = 'index'
  absolute = true
 }
```

But maybe it would be better to instantiate the runtime with some parameter.
Currently `\Neos\Fusion\Core\RuntimeFactory::createControllerContextFromEnvironment()` creates a HTTP request via `ServerRequest::fromGlobals()`. That means that absolute URLs will be generated with "http://localhost/...".

### Resources

resources will already work out of the box for regular HTTP requests even without configured baseUri because it will detect the base URI from the current server request.

For CLI requests one can use the already existing `baseUri` option of the resource targets to make that work out of the box:

So by replacing "_Resources/..." with "https://domain.tld/_Resources/..." in:

```yaml
Neos:
  Flow:
    resource:
      targets:
        localWebDirectoryStaticResourcesTarget:
          targetOptions:
            baseUri: '_Resources/Static/Packages/'

        localWebDirectoryPersistentResourcesTarget:
          targetOptions:
            baseUri: '_Resources/Persistent/'
```

For _Multi Site Installations_ that approach will work as well, but it has the (probably unexpected) side effect of resources being loaded from a different domain in some cases.

To work around that, we _could_ introduce some global concept (as Marc Henry suggested [above](https://github.com/neos/flow-development-collection/pull/2744#discussion_r1138872190)) but I would prefer not to go down that path (since it is basically the same as setting some superglobal variable, which is considered bad practice for obvious reasons).

An alternative could be some "context" parameter for the `ResourceManager::getPublicPersistentResourceUri()` method, i.e.:
```php
$context = ResourceTargetContext::fromRequest($httpRequest);
$uri = $this->resourceManager->getPublicPersistentResourceUri($resource, $context);
```

or – to prevent breaking changes – with some new `ContextAwareInterface` (or similar):
```php
$uri = $this->resourceManager
    ->withContext($context)
    ->getPublicPersistentResourceUri($resource);
```

Reply:
Thanks for the write down.

I agree that we don’t have to change anything further in this pr.

But after discussing the matter with @sorenmalling and @fcool we all came to the conclusion that the global concept still makes sense and is our best / safe bet.

i will prepare a draft pr for this were we can discuss further

Reply:
> But after discussing the matter with @sorenmalling and @fcool we all came to the conclusion that the global concept still makes sense and is our best / safe bet.

I respectfully disagree and IMO this "global thinking" introduced a lot of issues. At least the "global global" approach as in Settings.

Reply:
Hi, i just discussed this also with @mficzel and we found a minimal invasive solution: we will just introduce the enviroment setting `FLOW_BASE_URI`

so BaseUriProvider::getConfiguredBaseUriOrFallbackToCurrentRequest will just check for
1. FLOW_BASE_URI
2. deprecated setting `"Neos.Flow.http.baseUri"` 
3. base uri of $this->bootstrap->getActiveRequestHandler()


(the env can technically even be set in edge cases dynamically to cover my use case https://github.com/neos/flow-development-collection/pull/2744#discussion_r1138872190)

That way the BaseUriProvider could mostly just stay the same and we dont introduce any getter or setter (what we might regret in the future)

Also we have a nice replacement for the setting `"Neos.Flow.http.baseUri"` which will still be deprecated.

Reply:
Created as promised this follow up pr: https://github.com/neos/flow-development-collection/pull/3002

i hope youre all happy with this solution ...

Reply:
But how is this different than the setting, apart from making it harder for some folks to use the feature?

Reply:
I agree to @kitsunet .. let's discuss this in #3002